pkg/serviceaccounttoken/secret.go

```diff
@@ -111,40 +140,57 @@ func SecretTemplate(sa *v1.ServiceAccount) *v1.Secret {
 				ServiceAccountSecretLabel: sa.Name,
 			},
 		},
-		Type: v1.SecretTypeServiceAccountToken,
+		Type: corev1.SecretTypeServiceAccountToken,
 	}
 
 }
 
 // serviceAccountSecretPrefix returns the prefix that will be used to generate the secret for the given service account.
-func serviceAccountSecretPrefix(sa *v1.ServiceAccount) string {
+func serviceAccountSecretPrefix(sa *corev1.ServiceAccount) string {
 	return fmt.Sprintf("%s-token-", sa.Name)
 }
 
 // ServiceAccountSecret returns the secret for the given Service Account.
 // If there are more than one, it returns the first. Can return a nil secret
 // and a nil error if no secret is found
-func ServiceAccountSecret(ctx context.Context, sa *v1.ServiceAccount, secretLister secretLister, secretClient clientv1.SecretInterface) (*v1.Secret, error) {
+func ServiceAccountSecret(ctx context.Context, sa *corev1.ServiceAccount, secretLister secretLister, secretClient clientv1.SecretInterface) (*corev1.Secret, error) {
+	return serviceAccountSecret(ctx, sa, secretLister, secretClient, "")
+}
+
+func serviceAccountSecret(ctx context.Context, sa *corev1.ServiceAccount, secretLister secretLister, secretClient clientv1.SecretInterface, lockKey string) (*corev1.Secret, error) {
 	if sa == nil {
 		return nil, fmt.Errorf("cannot get secret for nil service account")
 	}
+
 	secrets, err := secretLister(sa.Namespace, labels.SelectorFromSet(map[string]string{
 		ServiceAccountSecretLabel: sa.Name,
 	}))
 	if err != nil {
 		return nil, fmt.Errorf("could not get secrets for service account: %w", err)
 	}
+
 	if len(secrets) < 1 {
 		return nil, nil
 	}
-	var result *v1.Secret
+
+	if len(secrets) > 1 && lockKey != "" {
```

Comment:
I might be missing something, but couldn't this deadlock with the calling function `createServiceAccountSecret`? Since that function locks before it calls this function, the lock will be held by the calling function. This function will wait on the lock to proceed, which it will never get because the calling function is waiting for this one to finish.

I definitely might be missing something, let me know if I am wrong.

Reply:
I had pondered this at the time...but here's my reasoning (and I've run the tests a lot, which while not perfect, were able to detect concurrency issues).

First call comes in...

```
EnsureSecretForServiceAccount
    serviceAccountSecret
        acquire lock (but only if there's more than one secret) ** this is really a lock around deletion **
            find secret ** and delete additional secrets **
        release lock
        return possible secret
    if serviceAccountSecret returned no secret
        createServiceAccountSecret
            use serviceAccountSecret to check for a secret
                (which will only lock if there's additional secrets, which should've been delete during the earlier lock)
        create the secret
    wait for the secret to be populated with the token
```

Second call comes in...

```
EnsureSecretForServiceAccount
    serviceAccountSecret
       _again_ acquire lock if there's more than one secret (which there could be because we're deleting them in the other Goroutine) but this will wait until the deletion lock is released
    if serviceAccountSecret returns no secret then as above, create the secret.
```

There is a race between using `serviceAccountSecret` to get a Secret (and getting `nil` back) and creating the secret, this is the tradeoff of reducing the scope of the lock.

Reply:
To be clear, very happy to walk through this, my analysis could easily be wrong too 😄

Reply:
I think the problem case comes in on this part:

```
if serviceAccountSecret returned no secret
        createServiceAccountSecret
            use serviceAccountSecret to check for a secret
                (which will only lock if there's additional secrets, which should've been delete during the earlier lock)
```

In this case, `createServiceAccountSecret` locks before calling `serviceAccountSecret`. I think that you are correct that `serviceAccountSecret` only locks if it finds more than one secret. However, I think that this is possible. Another rancher replica (or other third party application) could have created additional secrets in this time frame. I think that this would produce the deadlock. I think that you could replicate this in a test by having the reactor return 0 secrets on the first get call, but 2 secrets on the second get call.